pkg/operator/controller/dnsnameresolver/controller.go

```diff
@@ -0,0 +1,217 @@
+			config.OperandNamespace: {},
+		},
+		DefaultLabelSelector: labels.SelectorFromSet(labels.Set{
+			discoveryv1.LabelServiceName: config.ServiceName,
+		}),
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Initialize the reconciler.
+	reconciler := &reconciler{
+		config:               config,
+		dnsNameResolverCache: dnsNameResolverCache,
+		client:               mgr.GetClient(),
+		res:                  NewResolver(corednsEndpointsSliceCache, config.ServiceName),
+	}
+
+	// Create an unmanaged controller using the reconciler.
+	c, err := controller.NewUnmanaged(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Watch for the DNSNameResolver resources using dnsNameResolverCache.
+	if err := c.Watch(source.Kind(dnsNameResolverCache, &networkv1alpha1.DNSNameResolver{}), &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, nil, err
+	}
+
+	// Watch for the CoreDNS pod EndpointSlices using corednsEndpointsSliceCache.
+	if err := c.Watch(source.Kind(corednsEndpointsSliceCache, &discoveryv1.EndpointSlice{}), handler.EnqueueRequestsFromMapFunc(skipReconcileForEndpointSlice)); err != nil {
+		return nil, nil, err
+	}
+
+	return c, []cache.Cache{dnsNameResolverCache, corednsEndpointsSliceCache}, nil
+}
+
+// skipReconcileForEndpointSlice just skips the reconcile if any event related to the CoreDNS pod
+// endpointslices is received.
+func skipReconcileForEndpointSlice(context.Context, client.Object) []reconcile.Request {
+	return nil
+}
+
+type Config struct {
+	OperandNamespace         string
+	ServiceName              string
+	DNSNameResolverNamespace string
+}
+
+// reconciler handles the actual DNSNameResolver reconciliation logic in response to events.
+type reconciler struct {
+	config Config
+
+	dnsNameResolverCache cache.Cache
+	client               client.Client
+	res                  *Resolver
+	startResolver        sync.Once
+}
+
+// Reconcile expects request to refer to an DNSNameResolver resource, and will do all the work to
+// keep the status of the resource updated.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	// Ensure that the resolver is started before doing anything else.
+	r.startResolver.Do(func() {
+		r.res.Start()
+	})
+
+	logrus.Infof("reconciling request: %v", request)
+
+	// Get the DNSNameResolver resource.
+	dnsNameResolver := &networkv1alpha1.DNSNameResolver{}
+	if err := r.dnsNameResolverCache.Get(ctx, request.NamespacedName, dnsNameResolver); err != nil {
+
+		// Check if the DNSNameResolver resource is deleted. If so, delete DNS names matching the DNSNameResolver resource
+		// from the resolver.
+		if errors.IsNotFound(err) {
+			r.res.Delete(request.Name)
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	// Get a copy of the DNSNameResolver resource.
+	dnsNameResolver = dnsNameResolver.DeepCopy()
```

Comment:
Please name the copy something new.

Reply:
Moved to openshift/coredns-ocp-dnsnameresolver#2